src/v/storage/compaction_reducers.h

```diff
@@ -149,6 +151,11 @@ class copy_data_segment_reducer : public compaction_reducer {
     // Offset to keep in case the index is empty as of getting to this offset.
     model::offset _segment_last_offset;
     segment_appender* _appender;
+
+    // Compacted index writer for the newly written segment. May not be
+    // supplied if the compacted index isn't expected to change, e.g. when
+    // rewriting a single segment filtering with its own compacted index.
+    compacted_index_writer* _compacted_idx;
     index_state _idx;
```

Comment:
I see. We need to self compact first, so we still need to build the old compacted segment index. Still, I don't get the purpose of the new one.

Reply:
In general having thew new compacted index seems useful in cases where there are subsequent self compactions, to avoid having to rebuild it by reading the segment. That will happen after restarting, since the segment bitflags are in memory only.

Reply:
I'm still confused. What does `_compacted_idx` point to?

Reply:
I've re-read `deduplicate_segment` and it makes more sense now. Is the idea to update both the compaction index via (`compaction_index_writer`) and the segment index via the return value of `deduplicate_segment`?

Reply:
Right. FWIW I have a follow-up change to pass the `index_state` in as an input too, so the compacted index writer, index, and appender would all be decoupled from a single segment reducer, the idea there is that rather than rewriting a segment at a time, we could merge 2 (or more eventually) segments at a time by having the segment reducers share the writers.